### PR TITLE
Relation batcher

### PR DESCRIPTION
**Pull request by warrd**
RFC: https://github.com/wellcomecollection/docs/pull/42

### CODE REVIEW

File: pipeline/relation_embedder/batcher/src/main/scala/uk/ac/wellcome/platform/batcher/BatcherWorkerService.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+          .map(_.toList.unzip)
+          .mapAsync(1) { case (msgs, paths) => processPaths(msgs, paths) }
+          .flatMapConcat(identity)
+    )
+
+  private def processPaths(
+    msgs: List[SQSMessage],
+    paths: List[String]): Future[Source[SQSMessage, NotUsed]] =
+    generateSelectors(paths)
+      .groupedWithin(batchSize, batchFlushInterval)
```
Comment (warrd) [inline pipeline/relation_embedder/batcher/src/main/scala/uk/ac/wellcome/platform/batcher/BatcherWorkerService.scala:50]: Note for me: these should be grouped by root too

File: pipeline/relation_embedder/batcher/src/main/scala/uk/ac/wellcome/platform/batcher/Selector.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  * - The `Children` of some `path`
+  * - All `Descendents` of some `path`
+  */
+sealed trait Selector {
+
+  val path: String
+
+  /** Returns a list of all selectors which would also match this selector.
+    * We use these to filter out any unnecessary selectors when another less
+    * specific one already exists.
```
Comment (warrd) [inline pipeline/relation_embedder/batcher/src/main/scala/uk/ac/wellcome/platform/batcher/Selector.scala:15]: Note for me: "broader one"

File: pipeline/relation_embedder/batcher/src/main/scala/uk/ac/wellcome/platform/batcher/BatcherWorkerService.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      .map { failedIndices =>
+        val failedIdxSet = failedIndices.toSet
+        Source(msgs).zipWithIndex
+          .collect {
+            case (msg, idx) if !failedIdxSet.contains(idx) => msg
+          }
+      }
+
+  private def generateSelectors(
+    paths: List[String]): Source[(Selector, Long), NotUsed] = {
```
Comment (warrd) [inline pipeline/relation_embedder/batcher/src/main/scala/uk/ac/wellcome/platform/batcher/BatcherWorkerService.scala:74]: Note for me: add `info` logs to this process

**Review by alicefuzier — APPROVED** [review]
🎉

**Pull request merged**